src/util/hw_topo/mpir_hw_topo.c

```diff
@@ -0,0 +1,948 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#include "mpiimpl.h"
+#include "mpir_hw_topo.h"
+
+#ifdef HAVE_HWLOC
+#include "hwloc.h"
+#endif
+
+#ifdef HAVE_NETLOC
+#include "mpir_netloc.h"
+#endif
+
+/*
+ * Hardwre topology
+ */
+static struct {
+#ifdef HAVE_HWLOC
+    hwloc_topology_t hwloc_topology;
+    hwloc_cpuset_t bindset;
+#endif
+    int bindset_is_valid;
+#ifdef HAVE_NETLOC
+    netloc_topology_t netloc_topology;
+    MPIR_Netloc_network_attributes network_attr;
+#endif
+} hw_topo = {
+0};
+
+enum {
+    TO,
+    FROM
+};
```

Comment:
Even though this enum are only used locally, it still covers the scope of entire file, and make it a bit more verbose, e.g. add some prefix, may help the readability.

Reply:
These are only used as arguments within conversion functions, is it really necessary to prefix them?

Reply:
It is not really necessary, of course. But simple words should be avoided unless it is only used in a narrow scope when reviewers can see its context in one view.  Here, what does `TO/FROM` mean? Adding a few prefix or comment to it so one does't need search them to figure out their purposes would be nice.

Reply:
These directions are used with, e.g., `obj_type_to_from_hw_obj_type` and `TO/FROM` select the direction from which the conversion happens. `TO` means I give the function a `MPIR_NODE_OBJ_TYPE__` and I get a corresponding hardware object type `HWLOC_OBJ_TYPE_`. From means the opposite, I give the function an hwloc object type and I get an MPIR object type in return.

Reply:
I see. How about `MPIR_OBJ_TO_HWLOC` and `HWLOC_OBJ_TO_MPIR`?

Reply:
Instead of `obj_type_to_from_hw_obj_type`, how about `convert_hw_obj_type`?

Reply:
mmm this name seems to implicitly say that the function always takes an hw object type and converts it into something else. The initial name I was using was `mpir_obj_type_to_from_hw_obj_type` but that `mpir` prefix seemed like a namespace for a static function. `convert_obj_type` might be better as you can say what object type you want to convert using `MPIR_OBJ_TO_HWLOC` and `HWLOC_OBJ_TO_MPIR`?

Reply:
convert _node_obj_type?

Reply:
`convert_node_obj_type`, `convert_node_os_obj_type` and `convert_net_obj_type`?

Reply:
sounds good